--- ./src/flash/nor/at91sam3.c	2016-08-23 10:15:15.702541364 -0700
+++ ./src/flash/nor/at91sam3.c-new	2016-08-23 10:10:21.945084000 -0700
@@ -3246,7 +3246,7 @@
 /*   return r; */
 /* } */
 /* ======================================== */
-
+#if 0
 static const uint8_t
 	sam3_page_write_opcodes[] = {
 	/*  24 0000 0446                mov	r4, r0 */
@@ -3284,6 +3284,7 @@
 	0xfb, 0xd0,
 	0x00, 0xBE				/* bkpt #0 */
 };
+#endif //0
 
 static int sam3_page_write(struct sam3_bank_private *pPrivate, unsigned pagenum, uint8_t *buf)
 {
