Index: gcc-5.2.0/gcc/config/arm/t-arm-elf
===================================================================
--- a/gdb/gdbarch.c
+++ b/gdb/gdbarch.c
@@ -4976,6 +4976,8 @@ register_gdbarch_init (enum bfd_architec
 		       gdbarch_init_ftype *init)
 {
   gdbarch_register (bfd_architecture, init, NULL);
+  /* explicitly register bfd_arch_iamcu, will be treated as i386! */
+  gdbarch_register (bfd_arch_iamcu, init, NULL);
 }
 
 
