diff --git a/arch-init.c b/arch-init.c
--- a/arch_init.c	2015-01-22 08:19:03.000000000 -0800
+++ b.arch_init.c	2016-05-11 09:53:05.000000000 -0700
@@ -90,6 +90,8 @@
 #define QEMU_ARCH QEMU_ARCH_MIPS
 #elif defined(TARGET_MOXIE)
 #define QEMU_ARCH QEMU_ARCH_MOXIE
+#elif defined(TARGET_NIOS2)
+#define QEMU_ARCH QEMU_ARCH_NIOS2
 #elif defined(TARGET_OPENRISC)
 #define QEMU_ARCH QEMU_ARCH_OPENRISC
 #elif defined(TARGET_PPC)
diff --git a/include/sysemu/arch-init.h b/include/sysemu/arch-init.h
--- a/include/sysemu/arch_init.h 2015-01-22 08:19:05.000000000 -0800
+++ a/include/sysemu/arch_init.h 2016-05-11 09:55:01.000000000 -0700
@@ -22,6 +22,7 @@
     QEMU_ARCH_OPENRISC = 8192,
     QEMU_ARCH_UNICORE32 = 0x4000,
     QEMU_ARCH_MOXIE = 0x8000,
+    QEMU_ARCH_NIOS2 = 0x10000,
 };
