diff --git a/configure b/configure
--- a/configure	2015-01-22 08:19:03.000000000 -0800
+++ b/configure	2016-05-11 09:23:27.000000000 -0700
@@ -3000,7 +3006,7 @@
 fdt_required=no
 for target in $target_list; do
   case $target in
-    aarch64*-softmmu|arm*-softmmu|ppc*-softmmu|microblaze*-softmmu)
+    aarch64*-softmmu|arm*-softmmu|ppc*-softmmu|microblaze*-softmmu|nios2*-softmmu)
       fdt_required=yes
     ;;
   esac
@@ -4968,6 +4974,10 @@
   ;;
   moxie)
   ;;
+  nios2)
+    TARGET_ARCH=nios2
+    # bflt="yes"
+  ;;
   or32)
     TARGET_ARCH=openrisc
     TARGET_BASE_ARCH=openrisc
@@ -5167,6 +5177,10 @@
     echo "CONFIG_MOXIE_DIS=y"  >> $config_target_mak
     echo "CONFIG_MOXIE_DIS=y"  >> config-all-disas.mak
   ;;
+  nios2)
+    echo "CONFIG_NIOS2_DIS=y"  >> $config_target_mak
+    echo "CONFIG_NIOS2_DIS=y"  >> config-all-disas.mak
+  ;;
   or32)
     echo "CONFIG_OPENRISC_DIS=y"  >> $config_target_mak
     echo "CONFIG_OPENRISC_DIS=y"  >> config-all-disas.mak
