
diff --git a/hw/char/Makefile.objs b/hw/char/Makefile.objs
--- a/hw/char/Makefile.objs	2015-01-22 08:19:03.000000000 -0800
+++ b/hw/char/Makefile.objs	2016-05-11 16:25:00.000000000 -0700
@@ -6,6 +6,7 @@
 common-obj-$(CONFIG_SERIAL_PCI) += serial-pci.o
 common-obj-$(CONFIG_VIRTIO) += virtio-console.o
 common-obj-$(CONFIG_XILINX) += xilinx_uartlite.o
+common-obj-$(CONFIG_ALTERA) += altera_uart.o altera_juart.o
 common-obj-$(CONFIG_XEN_BACKEND) += xen_console.o
 common-obj-$(CONFIG_CADENCE) += cadence_uart.o
 
diff --git a/hw/timer/Makefile.objs b/hw/timer/Makefile.objs
--- a/hw/timer/Makefile.objs	2015-01-22 08:19:04.000000000 -0800
+++ b/hw/timer/Makefile.objs	2016-05-11 16:23:13.000000000 -0700
@@ -10,6 +10,7 @@
 common-obj-$(CONFIG_PUV3) += puv3_ost.o
 common-obj-$(CONFIG_TWL92230) += twl92230.o
 common-obj-$(CONFIG_XILINX) += xilinx_timer.o
+common-obj-$(CONFIG_ALTERA) += altera_timer.o
 common-obj-$(CONFIG_SLAVIO) += slavio_timer.o
 common-obj-$(CONFIG_ETRAXFS) += etraxfs_timer.o
 common-obj-$(CONFIG_GRLIB) += grlib_gptimer.o
