Index: Makefile.am
===================================================================
diff --git a/src/target/Makefile.am b/src/target/Makefile.am
--- a/src/target/Makefile.am	2016-04-19 10:27:02.380576802 -0700
+++ b/src/target/Makefile.am	2016-04-19 10:32:41.106256452 -0700
@@ -36,6 +36,7 @@
 	$(MIPS32_SRC) \
 	$(NDS32_SRC) \
 	$(INTEL_IA32_SRC) \
+	$(ARC32_SRC) \
 	avrt.c \
 	dsp563xx.c \
 	dsp563xx_once.c \
@@ -127,9 +128,22 @@
 
 INTEL_IA32_SRC = \
 	quark_x10xx.c \
+	quark_d20xx.c \
+	quark_d2000.c \
+	quark_se.c \
 	lakemont.c \
 	x86_32_common.c
 
+ARC32_SRC = \
+	arc_quark.c \
+	arc32.c \
+	arc_dbg.c \
+	arc_jtag.c \
+	arc_mem.c \
+	arc_mntr.c \
+	arc_ocd.c \
+	arc_regs.c
+
 noinst_HEADERS = \
 	algorithm.h \
 	arm.h \
@@ -199,7 +211,14 @@
 	nds32_v3m.h \
 	nds32_aice.h \
 	lakemont.h \
-	x86_32_common.h
+	x86_32_common.h \
+	arc32.h \
+	arc_dbg.h \
+	arc_jtag.h \
+	arc_mem.h \
+	arc_mntr.h \
+	arc_ocd.h \
+	arc_regs.h
 
 ocddatadir = $(pkglibdir)
 nobase_dist_ocddata_DATA =
