diff --git a/src/helper/options.c b/src/helper/options.c
index 5351e82..7cb3f84 100644
--- a/src/helper/options.c
+++ b/src/helper/options.c
@@ -95,8 +95,10 @@ static void add_default_dirs(void)
 	run_prefix = "";
 #endif
 
+#if 0 /* These messages can confuse users */
 	LOG_DEBUG("bindir=%s", BINDIR);
 	LOG_DEBUG("pkgdatadir=%s", PKGDATADIR);
+#endif
 	LOG_DEBUG("run_prefix=%s", run_prefix);
 
 	/*
@@ -124,7 +126,23 @@ static void add_default_dirs(void)
 		}
 	}
 #endif
+#if 1 /* ZEPHYR_SDK */
+	const char *zephyr_sdk = getenv("ZEPHYR_SDK_INSTALL_DIR");
 
+	if (zephyr_sdk) {
+		path = alloc_printf("%s/sysroots/i686-pokysdk-linux/usr/share/openocd/site", zephyr_sdk);
+		if (path) {
+			add_script_search_dir(path);
+			free(path);
+		}
+
+		path = alloc_printf("%s/sysroots/i686-pokysdk-linux/usr/share/openocd/scripts", zephyr_sdk);
+		if (path) {
+			add_script_search_dir(path);
+			free(path);
+		}
+	}
+#else
 	path = alloc_printf("%s%s%s", run_prefix, PKGDATADIR, "/site");
 	if (path) {
 		add_script_search_dir(path);
@@ -136,6 +154,8 @@ static void add_default_dirs(void)
 		add_script_search_dir(path);
 		free(path);
 	}
+#endif
+
 	add_script_search_dir("../scripts");
 	add_script_search_dir("scripts");
 }
